### PR TITLE
[Refactor] replace DistributionDisjoinSet by using common UnionFind class

### PR DESCRIPTION
**Pull request by packy92**
Fixes #issue
Fix problem similar in #34140

Ohter changes:
1. rename PropertyInfo to EquivalentDescriptor and change it to a private field only in `HashDistributionSpec` ensure all fields of other subclass of DistributionSpec to be immutable.
2. ensure all empty property use a singleton instance and immutable.
3. use UnionFind class to replace DistributionDisjoinSet

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java [hunk 1]
```diff
@@ -285,34 +286,27 @@ private HashDistributionSpec processHashDistributionSpec(
...
+        public void updateDistributionUnionFind(UnionFind<DistributionCol> unionFind) {
+            UnionFind<DistributionCol> tmp = unionFind.copy();
+            unionFind.clear();
+            for (Set<DistributionCol> distributionColSet : tmp.getAllGroups()) {
+                DistributionCol first = null;
+                for (DistributionCol col : distributionColSet) {
+                    if (first == null) {
+                        first = col;
+                    }
+                    unionFind.union(first, col);
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java:307]: Why not convert it by `columnMapping `?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/SortProperty.java [hunk 2]
```diff
@@ -28,8 +29,8 @@ public class SortProperty implements PhysicalProperty {
 
     public static final SortProperty EMPTY = new SortProperty();
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/SortProperty.java:30]: How to prevent changing `EMPTY`?

Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/SortProperty.java:30]: use a private constructor
```
private SortProperty() {
        this.spec = new OrderSpec(ImmutableList.of());
    }
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java (hunk 1 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/OptExpressionDuplicator.java]: has a probelem here. I fix it.

**Review by satanson — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**